src/modules/multiblade/clustering/ReferenceDataTest.cpp

```diff
@@ -0,0 +1,349 @@
+    uint32_t MatchedEvents{0}; // EFU - FP common events
+    uint32_t TimerWraps{0}; // Count when CAEN timer resets
+    uint32_t EventsInvalidWireGap{0}; // Events with wire gaps
+    uint32_t EventsInvalidStripGap{0}; // Events with strip gaps
+    uint32_t EventsInvalidThresh{0}; // Events with failed (wire) threshold
+  } Stats;
+
+  // Zero out stat counters
+  void clearStats() { std::fill_n((char*)&Stats, sizeof(Stats), 0); }
+
+  // Create a list of unmatched but 'good' events
+  // std::vector<Event> EventsExtra;
+
+  // Detect large jumps in readout time
+  void FixJumpsAndSort(std::vector<struct MBHits> & vec, bool sort);
+
+  // Do the clustering and event calulations
+  void LoadAndProcessReadouts(std::vector<struct MBHits> & vec);
+
+  // Compare previously calculated events with reference interpretation
+  // print out some stats
+  void CountMatches(std::vector<struct MBEvents> & evts, bool DiscardThresh);
+
+  bool thresholdCheck(uint8_t DigIndex, uint16_t GlobalChannel, uint16_t AdcValue);
+
+protected:
+  // builder uses a 2us time boxing
+  EventBuilder builder{2010};
+};
+
+bool compareByTime(const struct MBHits & a, const struct MBHits & b) {
+  return a.Time < b.Time;
+}
+
+bool isEqual(float t, uint16_t x, uint16_t y, struct MBEvents & res) {
+  return (fabsf(t - res.time) < 0.0000009) // 900ns
+          and ((uint16_t)(std::round(res.x)) == x)
+          and ((uint16_t)(std::round(res.y)) == y);
+}
+
+
+// MB18 specific
+bool isWire(uint16_t channel) {
+  return channel < NumberOfWires;
+}
+
+// MB18 specific - wires and strips share channel space, but we need each
+// coordinate to start at 0
+uint16_t GetWireCoord(uint32_t GlobalChannel) {
+  assert(isWire(GlobalChannel));
+  return GlobalChannel;
+}
+
+uint16_t GetStripCoord(uint32_t GlobalChannel) {
+  assert(GlobalChannel >= NumberOfWires);
+  return GlobalChannel - NumberOfWires;
+}
+
+bool ReferenceDataTest::thresholdCheck(uint8_t DigIndex, uint16_t GlobalChannel, uint16_t AdcValue) {
+  if (not isWire(GlobalChannel)) {
+    uint16_t StripCh = GlobalChannel;
+    if (AdcValue < 1) {
+      XTRACE(CLUSTER, DEB, "Strip threshold of 0 not met for ch %u ", StripCh);
+      return false;
+    } else {
+      return true;
+    }
+  } else { // wires
+    uint16_t WireCh = GetWireCoord(GlobalChannel);
+    uint16_t Thresh = builder.Thresholds[DigIndex][WireCh + 1];
+    if (AdcValue < Thresh) {
+      XTRACE(CLUSTER, DEB, "Wire threshold %u not met for ch %u (%u)",
+        Thresh, WireCh, AdcValue);
+      return false;
+    } else {
+      return true;
+    }
+  }
+}
+
+void ReferenceDataTest::FixJumpsAndSort(std::vector<struct MBHits> & vec, bool sort) {
+  int64_t Gap{43'000'000};
+  int64_t PrevTime{0xffffffffff};
+  std::vector<struct MBHits> temp;
+
+  clearStats();
+  for (auto & MBHit : vec) {
+    int64_t Time = (uint64_t)(MBHit.Time*1000000000ULL);
+
+    if ((PrevTime - Time ) < Gap) {
```

Comment:
Is the order of terms correct here? If I understand this correctly, with increasing times, the difference will always be negative and less than `Gap`.

Reply:
Time goes from 0 to some max value and then wraps so the diff can be both negative and positive